File: apps/mobile_android/test/readiness_test.dart
import 'package:flutter_test/flutter_test.dart';

import '../lib/readiness.dart';

void main() {
  group('computeReadiness', () {
    test('neutral inputs → 75 baseline, high band', () {
      final r = computeReadiness(const ReadinessInputs(tsb: 0));
      expect(r.score, 75);
      expect(r.band, ReadinessBand.high);
      expect(r.contributors, hasLength(1));
    });

    test('all inputs null → 75 baseline, no contributors', () {
      final r = computeReadiness(const ReadinessInputs(tsb: null));
      expect(r.score, 75);
      expect(r.contributors, isEmpty);
      expect(r.advice, matches(RegExp(r'good day|push the pace', caseSensitive: false)));
    });

    test('heavy fatigue + bad sleep → low band', () {
      final r = computeReadiness(
        const ReadinessInputs(tsb: -25, sleepHours: 4),
      );
      // -20 (TSB) + -25 (sleep) = -45 → 75-45 = 30 → low.
      expect(r.score, 30);
      expect(r.band, ReadinessBand.low);
      expect(r.advice, matches(RegExp('easy|rest', caseSensitive: false)));
    });

    test('fresh + great sleep → high band', () {
      final r = computeReadiness(
        const ReadinessInputs(
          tsb: 10,
          sleepHours: 8.5,
          restingHrBpm: 55,
          baselineRestingHrBpm: 58,
        ),
      );
      // 75 + 8 + 5 + 3 = 91.
      expect(r.score, 91);
      expect(r.band, ReadinessBand.high);
      expect(r.advice, matches(RegExp('harder effort')));
    });

    test('clamps to 0..100 on extreme inputs', () {
      final low = computeReadiness(
        const ReadinessInputs(
          tsb: -50,
          sleepHours: 2,
          restingHrBpm: 100,
          baselineRestingHrBpm: 55,
        ),
      );
      expect(low.score, 12);
      expect(low.band, ReadinessBand.low);

      final high = computeReadiness(
        const ReadinessInputs(
          tsb: 10,
          sleepHours: 8,
          restingHrBpm: 50,
          baselineRestingHrBpm: 55,
        ),
      );
      expect(high.score, inInclusiveRange(75, 100));
    });

    test('TSB > +25 (over-tapered) → small negative not positive', () {
      final r = computeReadiness(const ReadinessInputs(tsb: 30));
      expect(r.score, 75 - 3);
      final tsb = r.contributors
          .firstWhere((c) => c.kind == ReadinessContributorKind.form);
      expect(tsb.delta, -3);
      expect(tsb.note, matches(RegExp('Over-tapered')));
    });

    test('resting HR +12 above baseline → strong negative', () {
      final r = computeReadiness(
        const ReadinessInputs(
          tsb: 0,
          restingHrBpm: 70,
          baselineRestingHrBpm: 58,
        ),
      );
      expect(r.score, 75 - 18);
      expect(r.band, ReadinessBand.moderate);
      expect(r.advice,
          matches(RegExp('illness|under-recovery', caseSensitive: false)));
    });

    test('dominant input drives the advice line', () {
      final r = computeReadiness(
        const ReadinessInputs(tsb: -3, sleepHours: 4),
      );
      expect(r.advice,
          matches(RegExp('Very little sleep|compromised', caseSensitive: false)));
    });

    test('partial inputs — only sleep present', () {
      final r = computeReadiness(
        const ReadinessInputs(tsb: null, sleepHours: 8.5),
      );
      expect(r.score, 80);
      expect(r.band, ReadinessBand.high);
      expect(r.contributors, hasLength(1));
    });

    test('band thresholds', () {
      expect(
          computeReadiness(const ReadinessInputs(tsb: 0)).band,
          ReadinessBand.high);
      expect(
          computeReadiness(const ReadinessInputs(tsb: -10)).band,
          ReadinessBand.moderate);
      expect(
          computeReadiness(const ReadinessInputs(tsb: -25)).band,
          ReadinessBand.moderate);
      expect(
          computeReadiness(
                  const ReadinessInputs(tsb: -25, sleepHours: 3))
              .band,
          ReadinessBand.low);
    });

    test('score is deterministic — same inputs → same output', () {
      final a = computeReadiness(
          const ReadinessInputs(tsb: 5, sleepHours: 7));
      final b = computeReadiness(
          const ReadinessInputs(tsb: 5, sleepHours: 7));
      expect(a.score, b.score);
      expect(a.band, b.band);
      expect(a.advice, b.advice);
    });

    test('a tie in |delta| is broken by insertion order, not by the sort', () {
      // A TSB of -12 and six hours of sleep both score -12, and the two carry
      // DIFFERENT advice, so which one wins has to be a contract. Dart's
      // List.sort is unstable where JS's is stable, so this needs the
      // explicit index tiebreak to match the web twin.
      final r = computeReadiness(
          const ReadinessInputs(tsb: -12, sleepHours: 6));
      expect(r.contributors[0].delta, -12);
      expect(r.contributors[1].delta, -12);
      expect(r.advice, startsWith('Fatigued from recent training.'));
    });

    test('an all-zero tie still resolves to the first contributor', () {
      final r = computeReadiness(const ReadinessInputs(
        tsb: 0,
        sleepHours: 10,
        restingHrBpm: 50,
        baselineRestingHrBpm: 50,
      ));
      expect(r.contributors, hasLength(3));
      expect(r.contributors.map((c) => c.kind).toList(), [
        ReadinessContributorKind.form,
        ReadinessContributorKind.sleep,
        ReadinessContributorKind.restingHr,
      ]);
      expect(r.contributors.map((c) => c.delta).toList(), [0, 0, 0]);
      expect(r.advice, startsWith('Form is neutral.'));
    });

    test('the tiebreak does not override a genuinely larger contributor', () {
      // Sleep at -25 beats a TSB of -12 even though TSB was added first.
      final r = computeReadiness(
          const ReadinessInputs(tsb: -12, sleepHours: 4));
      expect(r.advice, startsWith('Very little sleep'));
    });
  });
}

File: apps/web/src/lib/training/readiness.test.ts
import { test } from 'node:test';
import { strict as assert } from 'node:assert';
import { computeReadiness } from './readiness';

test('neutral inputs → 75 baseline, high band', () => {
	const r = computeReadiness({ tsb: 0 });
	assert.equal(r.score, 75);
	assert.equal(r.band, 'high');
	assert.equal(r.contributors.length, 1);
});

test('all inputs null → 75 baseline, no contributors', () => {
	const r = computeReadiness({ tsb: null });
	assert.equal(r.score, 75);
	assert.equal(r.contributors.length, 0);
	// Advice is band-driven when no signals are available — 75 is the
	// "high" band, so we surface the encouraging variant.
	assert.match(r.advice, /good day|push the pace/i);
});

test('heavy fatigue + bad sleep → low band', () => {
	const r = computeReadiness({ tsb: -25, sleepHours: 4 });
	// -20 (TSB) + -25 (sleep) = -45 → 75 - 45 = 30 → low.
	assert.equal(r.score, 30);
	assert.equal(r.band, 'low');
	assert.match(r.advice, /easy|rest/i);
});

test('fresh + great sleep → high band', () => {
	const r = computeReadiness({
		tsb: 10,
		sleepHours: 8.5,
		restingHrBpm: 55,
		baselineRestingHrBpm: 58,
	});
	// 75 + 8 (fresh) + 5 (great sleep) + 3 (HR below baseline) = 91.
	assert.equal(r.score, 91);
	assert.equal(r.band, 'high');
	assert.match(r.advice, /harder effort/);
});

test('clamps to 0..100 on extreme inputs', () => {
	const low = computeReadiness({
		tsb: -50,
		sleepHours: 2,
		restingHrBpm: 100,
		baselineRestingHrBpm: 55,
	});
	assert.equal(low.score, 12);
	assert.equal(low.band, 'low');

	const high = computeReadiness({
		tsb: 10,
		sleepHours: 8,
		restingHrBpm: 50,
		baselineRestingHrBpm: 55,
	});
	assert.ok(high.score >= 75 && high.score <= 100);
});

test('TSB > +25 (over-tapered) → small negative not positive', () => {
	const r = computeReadiness({ tsb: 30 });
	assert.equal(r.score, 75 - 3);
	const tsb = r.contributors.find((c) => c.kind === 'form');
	assert.equal(tsb!.delta, -3);
	assert.match(tsb!.note, /Over-tapered/);
});

test('resting HR +10 above baseline → strong negative', () => {
	const r = computeReadiness({
		tsb: 0,
		restingHrBpm: 70,
		baselineRestingHrBpm: 58,
	});
	// 12 above → "Very above baseline" band, -18.
	assert.equal(r.score, 75 - 18);
	assert.equal(r.band, 'moderate');
	assert.match(r.advice, /illness|under-recovery/i);
});

test('contributor ordering: dominant input drives the advice line', () => {
	// Big sleep deficit dominates a mild TSB.
	const r = computeReadiness({ tsb: -3, sleepHours: 4 });
	// Sleep contributor (-25) beats TSB (-6).
	assert.match(r.advice, /Very little sleep|compromised/i);
});

test('partial inputs — only sleep present', () => {
	const r = computeReadiness({ tsb: null, sleepHours: 8.5 });
	// Baseline 75 + sleep +5 = 80, high band.
	assert.equal(r.score, 80);
	assert.equal(r.band, 'high');
	assert.equal(r.contributors.length, 1);
});

test('band thresholds: 40 = moderate boundary, 70 = high boundary', () => {
	// Neutral TSB → 75 → high band.
	assert.equal(computeReadiness({ tsb: 0 }).band, 'high');
	// Slight fatigue → 75-6=69 → moderate.
	assert.equal(computeReadiness({ tsb: -10 }).band, 'moderate');
	// Heavy fatigue → 75-20=55 → still moderate (>= 40).
	assert.equal(computeReadiness({ tsb: -25 }).band, 'moderate');
	// Heavy fatigue + bad sleep → 75-20-25=30 → low.
	assert.equal(computeReadiness({ tsb: -25, sleepHours: 3 }).band, 'low');
});

test('score is deterministic — same inputs → same output', () => {
	const a = computeReadiness({ tsb: 5, sleepHours: 7 });
	const b = computeReadiness({ tsb: 5, sleepHours: 7 });
	assert.equal(a.score, b.score);
	assert.equal(a.band, b.band);
	assert.equal(a.advice, b.advice);
});

test('a tie in |delta| is broken by insertion order, not by the sort', () => {
	// A TSB of -12 and six hours of sleep both score -12, and the two carry
	// DIFFERENT advice, so which one wins has to be a contract.
	// `Array.prototype.sort` is stable so the first contributor added wins;
	// the Dart twin's `List.sort` is NOT and carries an explicit index
	// tiebreak to match.
	const r = computeReadiness({ tsb: -12, sleepHours: 6 });
	assert.equal(r.contributors[0].delta, -12);
	assert.equal(r.contributors[1].delta, -12);
	assert.match(r.advice, /^Fatigued from recent training\./);
});

test('an all-zero tie still resolves to the first contributor', () => {
	const r = computeReadiness({
		tsb: 0,
		sleepHours: 10,
		restingHrBpm: 50,
		baselineRestingHrBpm: 50,
	});
	assert.equal(r.contributors.length, 3);
	assert.deepEqual(
		r.contributors.map((c) => c.kind),
		['form', 'sleep', 'resting_hr'],
	);
	assert.deepEqual(
		r.contributors.map((c) => c.delta),
		[0, 0, 0],
	);
	assert.match(r.advice, /^Form is neutral\./);
});

test('the tiebreak does not override a genuinely larger contributor', () => {
	// Sleep at -25 beats a TSB of -12 even though TSB was added first.
	const r = computeReadiness({ tsb: -12, sleepHours: 4 });
	assert.match(r.advice, /^Very little sleep/);
});
